File: Core/rag_models.py
"""
Core/rag_models.py — Pydantic validation models for RAG data pipeline.

Provides validated dataclasses for:
  - ChunkPayload: what gets stored in Qdrant per chunk
  - RAGSearchResult: what hybrid_search() returns per result
  - QueryRewriteResult: multi-query expansion output
  - EvalSample: evaluation harness sample

Ported from ZEN_RAG.
"""

from __future__ import annotations
from typing import Any, List, Optional
from pydantic import BaseModel, ConfigDict, Field, field_validator


class ChunkPayload(BaseModel):
    """Validated payload stored in Qdrant for each chunk."""

    model_config = ConfigDict(extra="allow")

    text: str = Field(..., min_length=1, description="Chunk text content")
    url: Optional[str] = Field(None, description="Source URL or file path")
    title: Optional[str] = Field(None, description="Document title")
    scan_root: Optional[str] = Field(None, description="Root path scanned")
    chunk_index: int = Field(0, ge=0, description="Position of this chunk in the document")
    is_table: bool = Field(False, description="True if chunk represents tabular data")
    sheet_name: Optional[str] = Field(None, description="Excel sheet name (tables only)")
    parent_id: Optional[str] = Field(None, description="Parent chunk hash (hierarchical chunking)")
    doc_type: Optional[str] = Field(None, description="Document type: pdf, docx, web, excel, ...")

    @field_validator("text")
    @classmethod
    def text_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Chunk text must not be blank")
        return v.strip()


class RAGSearchResult(BaseModel):
    """Validated search result returned by hybrid_search / search."""

    model_config = ConfigDict(extra="allow")

    text: str = Field(..., min_length=1, description="Retrieved chunk text")
    url: Optional[str] = Field(None, description="Source URL or file path")
    title: Optional[str] = Field(None, description="Document title")
    score: float = Field(0.0, description="Retrieval score (Qdrant cosine or fusion)")
    rerank_score: Optional[float] = Field(None, description="Post-reranking score")
    fusion_score: Optional[float] = Field(None, description="RRF fusion score")
    is_cached: bool = Field(False, description="True if result came from cache")
    parent_text: Optional[str] = Field(None, description="Parent chunk text for context window")
    is_table: bool = Field(False, description="True if chunk is tabular data")

    @field_validator("text")
    @classmethod
    def text_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("SearchResult text must not be blank")
        return v.strip()

    @field_validator("score", "rerank_score", "fusion_score", mode="before")
    @classmethod
    def coerce_float(cls, v: Any) -> Optional[float]:
        if v is None:
            return None
        try:
            return float(v)
        except (TypeError, ValueError):
            return 0.0


class QueryRewriteResult(BaseModel):
    """Output of multi-query expansion."""

    original: str = Field(..., description="Original user query")
    rewrites: List[str] = Field(default_factory=list, description="LLM-generated query variations")
    strategy: str = Field("passthrough", description="Strategy used: llm, template, passthrough")

    @property
    def all_queries(self) -> List[str]:
        """Original + all rewrites (deduplicated)."""
        seen = {self.original.lower()}
        result = [self.original]
        for q in self.rewrites:
            if q.lower() not in seen:
                seen.add(q.lower())
                result.append(q)
        return result


class EvalSample(BaseModel):
    """A single evaluation sample with ground truth."""

    query: str
    expected_answer: str
    retrieved_texts: List[str] = Field(default_factory=list)
    generated_answer: Optional[str] = None
    relevance_scores: List[float] = Field(default_factory=list)
    ndcg: Optional[float] = None
    mrr: Optional[float] = None

File: rust_output/src/rag_models.rs
/// Core/rag_models::py — Pydantic validation models for RAG data pipeline.
/// 
/// Provides validated dataclasses for:
/// - ChunkPayload: what gets stored in Qdrant per chunk
/// - RAGSearchResult: what hybrid_search() returns per result
/// - QueryRewriteResult: multi-query expansion output
/// - EvalSample: evaluation harness sample
/// 
/// Ported from ZEN_RAG.

use anyhow::{Result, Context};
use std::collections::HashSet;

/// Validated payload stored in Qdrant for each chunk.
#[derive(Debug, Clone)]
pub struct ChunkPayload {
    pub text: String,
    pub url: Option<String>,
    pub title: Option<String>,
    pub scan_root: Option<String>,
    pub chunk_index: i64,
    pub is_table: bool,
    pub sheet_name: Option<String>,
    pub parent_id: Option<String>,
    pub doc_type: Option<String>,
}

impl ChunkPayload {
    pub fn text_not_blank(v: String) -> Result<String> {
        if !v.trim().to_string() {
            return Err(anyhow::anyhow!("ValueError('Chunk text must not be blank')"));
        }
        Ok(v.trim().to_string())
    }
}

/// Validated search result returned by hybrid_search / search.
#[derive(Debug, Clone)]
pub struct RAGSearchResult {
    pub text: String,
    pub url: Option<String>,
    pub title: Option<String>,
    pub score: f64,
    pub rerank_score: Option<f64>,
    pub fusion_score: Option<f64>,
    pub is_cached: bool,
    pub parent_text: Option<String>,
    pub is_table: bool,
}

impl RAGSearchResult {
    pub fn text_not_blank(v: String) -> Result<String> {
        if !v.trim().to_string() {
            return Err(anyhow::anyhow!("ValueError('SearchResult text must not be blank')"));
        }
        Ok(v.trim().to_string())
    }
    pub fn coerce_float(v: Box<dyn std::any::Any>) -> Result<Option<f64>> {
        if v.is_none() {
            None
        }
        // try:
        {
            v.to_string().parse::<f64>().unwrap_or(0.0)
        }
        // except (TypeError, ValueError) as _e:
    }
}

/// Output of multi-query expansion.
#[derive(Debug, Clone)]
pub struct QueryRewriteResult {
    pub original: String,
    pub rewrites: Vec<String>,
    pub strategy: String,
}

impl QueryRewriteResult {
    /// Original + all rewrites (deduplicated).
    pub fn all_queries(&self) -> Vec<String> {
        // Original + all rewrites (deduplicated).
        let mut seen = HashSet::from([self.original.to_lowercase()]);
        let mut result = vec![self.original];
        for q in self.rewrites.iter() {
            if !seen.contains(&q.to_lowercase()) {
                seen.insert(q.to_lowercase());
                result.push(q);
            }
        }
        result
    }
}

/// A single evaluation sample with ground truth.
#[derive(Debug, Clone)]
pub struct EvalSample {
    pub query: String,
    pub expected_answer: String,
    pub retrieved_texts: Vec<String>,
    pub generated_answer: Option<String>,
    pub relevance_scores: Vec<f64>,
    pub ndcg: Option<f64>,
    pub mrr: Option<f64>,
}
